include/vicon/udp.h

```diff
@@ -228,28 +242,52 @@ class ObjectReference
 template<typename ObjectDataType = ObjectData>
 class UDPClient
 {
-public:
-    UDPClient() {}
+private:
+    /*
+     * NB: The first member of this struct *should* be object ID, but the
+     * Vicon system seems to just give a value of zero for every item, which
+     * isn't terribly helpful. So instead we distinguish objects based on their
+     * names.
+     */
+#ifdef _MSVC
+#pragma pack(push, 1)
+    struct RawObjectData {
+        uint8_t unused;
+        uint16_t itemDataSize;
+        char objectName[24];
+        double position[3], attitude[3];
+    };
+    #pragma pack(pop)
+#else
+    struct __attribute__((__packed__)) RawObjectData {
+        uint8_t unused;
+        uint16_t itemDataSize;
+        char objectName[24];
+        std::array<double, 3> position, attitude;
+    };
+#endif
 
+public:
+    UDPClient() = default;
     UDPClient(uint16_t port)
     {
         connect(port);
     }
 
     virtual ~UDPClient()
     {
-        // Set quit flag and join read thread
-        if(m_ReadThread.joinable()) {
-            m_ShouldQuit = true;
-            m_ReadThread.join();
-        }
+        // Set quit flag
+        m_ShouldQuit = true;
     }
 
     //----------------------------------------------------------------------------
     // Public API
     //----------------------------------------------------------------------------
     void connect(uint16_t port)
     {
+        // Lock until we have at least one packet from Vicon system
+        m_ConnectionMutex.lock();
```

Comment:
This seems a bit of an abuse of a mutex - I think ``std::condition_variable`` might be better here

Reply:
I don't particularly understand condvars too well... but don't you need a mutex for those anyway?

Reply:
Very good point :smile: